packages/subcommands/src/lib/Subcommand.ts

```diff
@@ -0,0 +1,189 @@
+	type ChatInputCommand,
+	UserError,
+	Identifiers,
+	type Awaitable,
+	type MessageCommand
+} from '@sapphire/framework';
+import type { Message } from 'discord.js';
+import {
+	ChatInputSubcommandMappings,
+	ChatInputSubcommandGroupMappings,
+	ChatInputSubcommandMappingValue,
+	ChatInputSubcommandToProperty,
+	MessageSubcommandMappings,
+	MessageSubcommandMappingValue,
+	MessageSubcommandToProperty,
+	type SubcommandMappingsArray
+} from './SubcommandMappings';
+import { type ChatInputSubcommandAcceptedPayload, Events, type MessageSubcommandAcceptedPayload } from './types/Events';
+
+export class SubCommandPluginCommand extends Command {
+	public readonly subcommands: SubcommandMappingsArray;
+
+	public constructor(context: PieceContext, options: SubcommandPluginCommandOptions) {
+		super(context, options);
+		this.subcommands = options.subcommands ?? [];
+	}
+
+	public messageRun(message: Message, args: Args, context: MessageCommand.RunContext): Awaitable<unknown> {
+		args.save();
+		const value = args.nextMaybe();
+		let defaultCommmand: MessageSubcommandMappingValue | null = null;
+
+		for (const mapping of this.subcommands) {
+			if (!(mapping instanceof MessageSubcommandMappings)) continue;
+			defaultCommmand = mapping.subcommands.find((s) => s.default === true) ?? null;
+			const subcommand = mapping.subcommands.find(({ name }) => name === value.value);
+			if (subcommand) return this.#handleMessageRun(message, args, context, subcommand);
+		}
+
+		// No subcommand matched, let's restore and try to run default, if any:
+		args.restore();
+		if (defaultCommmand) return this.#handleMessageRun(message, args, context, defaultCommmand);
+
+		// No match and no subcommand, return an err:
+		throw new UserError({ identifier: Identifiers.MessageSubcommandNoMatch, context });
+	}
+
+	public chatInputRun(interaction: ChatInputCommand.Interaction, context: ChatInputCommand.RunContext): Awaitable<unknown> {
+		const subcommandName = interaction.options.getSubcommand(false);
+		const subcommandGroupName = interaction.options.getSubcommandGroup(false);
+
+		if (subcommandName && !subcommandGroupName) {
+			for (const mapping of this.subcommands) {
+				if (!(mapping instanceof ChatInputSubcommandMappings)) continue;
+
+				const subcommand = mapping.subcommands.find(({ name }) => name === subcommandName);
+				if (subcommand) return this.#handleInteractionRun(interaction, context, subcommand);
+			}
+		}
+
+		if (subcommandGroupName) {
+			for (const mapping of this.subcommands) {
+				if (!(mapping instanceof ChatInputSubcommandGroupMappings)) continue;
+				if (mapping.groupName !== subcommandGroupName) continue;
+
+				const subcommand = mapping.subcommands.find(({ name }) => name === subcommandName);
+				if (subcommand) return this.#handleInteractionRun(interaction, context, subcommand);
+			}
+		}
+
+		// No match and no subcommand, return an err:
+		throw new UserError({ identifier: Identifiers.ChatInputSubcommandNoMatch, context });
```

Comment:
This is a hard error, please mark the method as async instead!

Reply:
Wdym?

Reply:
wouldn't this get caught by the fromAsync?

Reply:
No, it is at the end of the messageRun and chatInputRun methods, and is executed when no subcommand with the conditions applied above the throw is found